File: 2021/10/01/main.go
package main

func longestCommonSubsequence(text1 string, text2 string) int {
	m, n := len(text1), len(text2)
	a, b := []byte(text1), []byte(text2)

	cur := make([]int, m+1)
	prev := make([]int, m+1)

	for j := 0; j < n; j++ {
		for i := 0; i < m; i++ {
			rec := max(prev[i+1], cur[i])
			if a[i] == b[j] {
				cur[i+1] = max(rec, prev[i]+1)
			} else {
				cur[i+1] = rec
			}
		}
		cur, prev = prev, cur
	}
	return prev[m]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 2021/10/01/main.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        m, n = len(text1), len(text2)
        cur = [0]*(m+1)
        prev = cur
        for j in range(n):
            for i in range(m):
                rec = max(prev[i+1], cur[i])
                if text1[i] == text2[j]:
                    cur[i+1] = max(rec, prev[i]+1)
                else:
                    cur[i+1] = rec
            cur, prev = prev, cur
        return prev[m]